File: frontend/src/shared/i18n.ts
// Every user-visible string comes from translations/panel/<lang>.json, fetched
// from the backend over foyer/translations. Components never contain text.
import type { HomeAssistant } from "./types";

export type Strings = Record<string, unknown>;

const cache = new Map<string, Promise<Strings>>();

export function loadStrings(hass: HomeAssistant): Promise<Strings> {
  const language = hass.language;
  let pending = cache.get(language);
  if (!pending) {
    pending = hass
      .callWS<{ language: string; strings: Strings }>({
        type: "foyer/translations",
        language,
      })
      .then((result) => result.strings);
    // A failed load must be retried next time, not cached forever.
    pending.catch(() => cache.delete(language));
    cache.set(language, pending);
  }
  return pending;
}

/** Look up a dotted key and fill {placeholders}. A missing key renders as the key. */
export function t(
  strings: Strings | undefined,
  key: string,
  params: Record<string, string | number> = {},
): string {
  let node: unknown = strings;
  for (const part of key.split(".")) {
    if (node && typeof node === "object" && part in node) {
      node = (node as Record<string, unknown>)[part];
    } else {
      return key;
    }
  }
  if (typeof node !== "string") return key;
  return node.replace(/\{(\w+)\}/g, (match, name: string) =>
    name in params ? String(params[name]) : match,
  );
}

File: custom_components/foyer/i18n.py
"""Frontend and notification strings from ``translations/panel/<lang>.json``.

Home Assistant's own ``translations/<lang>.json`` is validated by hassfest
against a closed schema that has no room for panel text or ``help.<page>``, so
those strings live alongside it in ``translations/panel/``. Home Assistant
ignores the subdirectory; Foyer serves it over ``foyer/translations``.
"""

from __future__ import annotations

from functools import lru_cache
import json
from pathlib import Path
from typing import Any

PANEL_TRANSLATIONS = Path(__file__).parent / "translations" / "panel"
FALLBACK_LANGUAGE = "en"


def available_languages() -> list[str]:
    return sorted(p.stem for p in PANEL_TRANSLATIONS.glob("*.json"))


def resolve_language(language: str | None) -> str:
    """Map a Home Assistant language ("it", "en-GB", "pt-BR") to a file we have."""
    languages = set(available_languages())
    if language:
        for candidate in (language, language.split("-")[0]):
            if candidate in languages:
                return candidate
    return FALLBACK_LANGUAGE


@lru_cache(maxsize=8)
def load_strings(language: str) -> dict[str, Any]:
    """Blocking read; call from an executor. Cached for the process lifetime."""
    path = PANEL_TRANSLATIONS / f"{resolve_language(language)}.json"
    return json.loads(path.read_text(encoding="utf-8"))


def languages() -> list[dict[str, str]]:
    """Every language on disk, each named in itself. Blocking; use an executor.

    The name comes from the language's own file (``language.self``) rather than
    from a table here or in the frontend, so that adding a language is copying
    two files and nothing else (§20.3): a list written in code is a line a
    translator has to find, and one they will not know exists.
    """
    return [
        {"code": code, "name": translate(load_strings(code), "language.self")}
        for code in available_languages()
    ]


def translate(strings: dict[str, Any], key: str, **placeholders: str) -> str:
    """Look up a dotted key and fill ``{placeholders}``. Missing keys return the key."""
    node: Any = strings
    for part in key.split("."):
        if not isinstance(node, dict) or part not in node:
            return key
        node = node[part]
    if not isinstance(node, str):
        return key
    try:
        return node.format(**placeholders)
    except (KeyError, IndexError):
        return node
